### PR TITLE
Use soft-update to maintain document history

### PR DESCRIPTION
**Pull request by dnhatn**
Today we can use the soft-update feature from Lucene to maintain a history of document. This change simply replaces hard-update in the Engine by soft-update methods. Stale operations, delete, and no-ops will be handled in subsequent changes. This change is just a cut-over from hard-update to soft-update, no new functionality has been introduced.

### CODE REVIEW

**Comment by dnhatn** [comment]
test this please.

**Comment by dnhatn** [comment]
please test this.

**Comment by s1monw** [comment]
@elasticmachine test this please

**Review by s1monw — CHANGES_REQUESTED** [review]
looks great, I left some comments

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -4555,4 +4599,50 @@ void assertLuceneOperations(InternalEngine engine, long expectedAppends, long ex
...
+    /**
+     * A reader that does not exclude soft-deleted documents.
+     */
+    static final class UnfilteredReader extends FilterDirectoryReader {
+        static final class UnfilteredSubReaderWrapper extends SubReaderWrapper {
+            @Override
+            public LeafReader wrap(LeafReader in) {
+                return new FilterLeafReader(in) {
+                    @Override
+                    public CacheHelper getCoreCacheHelper() {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I removed the test and this wrapper in the PR, and will add it to Lucene.java later. I prefer to keep this change as a cut-over.

**Comment by dnhatn** [comment]
@s1monw I have addressed your comments. Can you please take a look? Thank you.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -1073,9 +1070,9 @@ private boolean assertDocDoesNotExist(final Index index, final boolean allowDele
     private void updateDocs(final Term uid, final List<ParseContext.Document> docs, final IndexWriter indexWriter) throws IOException {
         if (softDeleteEnabled) {
             if (docs.size() > 1) {
-                indexWriter.softUpdateDocuments(uid, docs, Lucene.getSoftDeleteDVMarker());
+                indexWriter.softUpdateDocuments(uid, docs, Lucene.newSoftDeleteField());
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think we can cache this in here in a constant so we don't need to re-create a new one every time.

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks @s1monw for reviewing.

**Pull request merged**